crypto/fipsmodule/rand/snapsafe_detect_test.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/base.h>
+#include <cstdint>
+
+#include <gtest/gtest.h>
+#include "snapsafe_detect.h"
+
+#if defined(OPENSSL_LINUX) && defined(AWSLC_SNAPSAFE_TESTING)
+#include <fcntl.h>
+#include <cstring>
+#include <sys/mman.h>
+
+#define NUMBER_OF_TEST_VALUES 5
+
+typedef struct sgn_test_s {
+  void* addr;
+  size_t pgsize;
+} sgn_test_s;
+
+int init_sgn_file(sgn_test_s* sgn_test);
+int init_sgn_file(sgn_test_s* sgn_test) {
+  const char* sgc_file_path = AWSLC_SYSGENID_PATH;
+  int fd_sgn = open(sgc_file_path, O_CREAT | O_RDWR | O_APPEND, (mode_t)0600);
+  if (fd_sgn == -1) {
+    return 0;
+  }
+  if (0 != lseek(fd_sgn, 0, SEEK_SET)) {
+    return 0;
+  }
+  
+  unsigned int set_sgn = 0;
+  if(0 >= write(fd_sgn, &set_sgn, sizeof(unsigned int))) {
+    return 0;
+  }
+  long page_size = sysconf(_SC_PAGESIZE);
+  if (page_size <= 0) {
+    return 0;
+  }
+  size_t pgsize = (size_t)page_size;
+
+  void* addr = mmap(NULL, pgsize, PROT_WRITE, MAP_SHARED, fd_sgn, 0);
+  if (addr == MAP_FAILED) {
+    return 0;
+  }
+
+  close(fd_sgn);
+
+  sgn_test->addr = addr;
+  sgn_test->pgsize = pgsize;
+
+  return 1;  
+}
+
+int set_sgn(const sgn_test_s* sgn_test, unsigned int val);
+int set_sgn(const sgn_test_s* sgn_test, unsigned int val) {
+  memcpy(sgn_test->addr, &val, sizeof(unsigned int));
+  if(0 != msync(sgn_test->addr, sgn_test->pgsize, MS_SYNC)) {
+    return 0;
+  }
+  return 1;
+}
+
+TEST(SnapsafeGenerationTest, SysGenIDretrievalTesting) {
+  sgn_test_s sgn_test;
+  if(1 != init_sgn_file(&sgn_test)) {
+    FAIL();
+  }
```

Comment:
ASSERTS will fail and stop executing the test and is our normal test convention.
```suggestion
  ASSERT_TRUE(init_sgn_file(&sgn_test)));
```